### ISSUE TITLE
Fixed type for atttribute

### ISSUE THREAD

**Issue by RainbowC0** [astral-sh/ty#4328]
### Summary

Type of an attribute is fixed from the initial deduction, which results in unexpected type errors once that gets modified or reassigned.

```python
class Attn_Net(nn.Module):

    def __init__(self, L = 1024, D = 256, dropout = False, n_classes = 1):
        super(Attn_Net, self).__init__()
        self.module = [
            nn.Linear(L, D),
            nn.Tanh()] # fixed type list[Linear|Tanh]
        if dropout:
            self.module.append(nn.Dropout(0.25))
            # Argument to bound method `list.append` is incorrect: Expected `Linear | Tanh`, 
            # found `Dropout`ty[invalid-argument-type](https://ty.dev/rules#invalid-argument-type)
        self.module.append(nn.Linear(D, n_classes))
        self.module = nn.Sequential(*self.module)

    def forward(self, x):
        return self.module(x), x
        # Object of type `list[Linear | Tanh]` is not callable
        # info: Union variant `list[Linear | Tanh]` is incompatible with this call site
        # info: Attempted to call union type `list[Linear | Tanh] | Sequential`ty[call-non-callable](https://ty.dev/rules#call-non-callable)
```

### Version

ty 0.0.73 (4bd2833c4 2026-08-18)

**Comment by carljm** [comment]
Thanks for the report!

There are kind of two separate issues shown in this example. The first can be minimized as:

```py
class C:
    def __init__(self):
        self.items = [1]
        self.items.append("a")  # Expected int, found Literal["a"]
```

And the second can be minimized as:

```py
class C:
    def __init__(self):
        self.value = 1
        self.value = "a"

    def get(self) -> str:
        return self.value  # Expected str, found int | str
```

Most important context here is that _every Python type checker_ currently errors in exactly the same way, for exactly the same reason, on both of these examples. And these limitations are both somewhat difficult to fix in a way that doesn't give inaccurate results on other code.

This means that it's not going to be a near-term priority to change our behavior in either of these cases, and (if you want to use a Python type checker on your code at all) your best bet will be to adjust the code slightly so it's more compatible with static analysis. Fortunately, that's quite easy to do: use a local variable (e.g. `layers`) instead of using `self.module`, for all the code up until the final assignment to `self.module`:

```py
class Attn_Net(nn.Module):

    def __init__(self, L = 1024, D = 256, dropout = False, n_classes = 1):
        super(Attn_Net, self).__init__()
        layers = [
            nn.Linear(L, D),
            nn.Tanh()] # fixed type list[Linear|Tanh]
        if dropout:
            layers.append(nn.Dropout(0.25))
        layers.append(nn.Linear(D, n_classes))
        self.module = nn.Sequential(*layers)

    def forward(self, x):
        return self.module(x), x
        # Object of type `list[Linear | Tanh]` is not callable
        # info: Union variant `list[Linear | Tanh]` is incompatible with this call site
        # info: Attempted to call union type `list[Linear | Tanh] | Sequential`ty[call-non-callable](https://ty.dev/rules#call-non-callable)
```

This makes ty happy with the code, and arguably makes the code clearer for a human reader, too, since it clarifies what type `self.module` is actually expected to be.

I'm going to close this as "not planned" for now, since it's combining two issues in one, and neither issue is on our immediate roadmap. If you feel strongly about one of the two examples above, and would like us to fix it at some point in the future, feel free to file a focused issue for it, and we can keep it open as a wishlist item.
